IdentityCore/src/cache/MSIDToken.m

```diff
@@ -22,7 +22,130 @@
 // THE SOFTWARE.
 
 #import "MSIDToken.h"
+#import "MSIDUserInformation.h"
 
 @implementation MSIDToken
 
+- (BOOL)isEqualToToken:(MSIDToken *)token
+{
+    if (!token)
+    {
+        return NO;
+    }
+    
+    BOOL result = YES;
+    result &= (!self.token && !token.token) || [self.token isEqualToString:token.token];
+    result &= (!self.idToken && !token.idToken) || [self.idToken isEqualToString:token.idToken];
+    result &= (!self.expiresOn && !token.expiresOn) || [self.expiresOn isEqualToDate:token.expiresOn];
+    result &= (!self.familyId && !token.familyId) || [self.familyId isEqualToString:token.familyId];
+    result &= (!self.clientInfo && !token.clientInfo) || [self.clientInfo isEqualToDictionary:token.clientInfo];
+    result &= (!self.additionalServerInfo && !token.additionalServerInfo) || [self.additionalServerInfo isEqualToDictionary:token.additionalServerInfo];
+    result &= self.tokenType == token.tokenType;
+    result &= (!self.resource && !token.resource) || [self.resource isEqualToString:token.resource];
+    result &= (!self.authority && !token.authority) || [self.authority isEqualToString:token.authority];
+    result &= (!self.clientId && !token.clientId) || [self.clientId isEqualToString:token.clientId];
+    
+    return result;
+}
+
+#pragma mark - NSObject
+
+- (BOOL)isEqual:(id)object
+{
+    if (self == object)
+    {
+        return YES;
+    }
+    
+    if (![object isKindOfClass:MSIDToken.class])
+    {
+        return NO;
+    }
+    
+    return [self isEqualToToken:(MSIDToken *)object];
+}
+
+- (NSUInteger)hash
+{
+    NSUInteger hash = self.token.hash;
+    hash ^= self.idToken.hash;
+    hash ^= self.expiresOn.hash;
+    hash ^= self.familyId.hash;
+    hash ^= self.clientInfo.hash;
+    hash ^= self.additionalServerInfo.hash;
+    hash ^= self.tokenType;
+    hash ^= self.resource.hash;
+    hash ^= self.authority.hash;
+    hash ^= self.clientId.hash;
+    
+    return hash;
+}
+
+#pragma mark - NSSecureCoding
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    _familyId = [coder decodeObjectOfClass:[NSString class] forKey:@"familyId"];
+    _expiresOn = [coder decodeObjectOfClass:[NSDate class] forKey:@"expiresOn"];
+    
+    NSString *accessToken = [coder decodeObjectOfClass:[NSString class] forKey:@"accessToken"];
+    NSString *refreshToken = [coder decodeObjectOfClass:[NSString class] forKey:@"refreshToken"];
+    
+    if (refreshToken)
+    {
+        _token = refreshToken;
+        _tokenType = MSIDTokenTypeRefreshToken;
+    }
+    else
+    {
+        _token = accessToken;
+        _tokenType = MSIDTokenTypeAccessToken;
+    }
+    
+    _additionalServerInfo = [coder decodeObjectOfClass:[NSDictionary class] forKey:@"additionalServer"];
+    _clientInfo = [coder decodeObjectOfClass:[NSMutableDictionary class] forKey:@"additionalClient"];
```

Comment:
clientInfo is not additionalClient. additionalClient is just additional properties to store with the token and not used by ADAL. clientInfo is additional data returned by the STS that has unique user identifier.

Reply:
fixed